template/template.go

```diff
@@ -296,3 +312,196 @@ func (e *executer) resolveOptions(options []ast.Option) (map[string]any, error)
+		rv := Selector{k, v}
+		if e.input[k] != "" {
+			res = append(res, rv)
+		} else {
+			return nil, fmt.Errorf("%w '%s'", ErrSelection, res)
+		}
+	}
+
+	return res, nil
+}
+
+func (e *executer) resolvePreferences(m ast.Matcher, res []Selector) [][]string {
+	// Step 2: Resolve Preferences
+	var pref [][]string //nolint:prealloc
+
+	for i := range res {
+		var keys []string
+
+		for _, variant := range m.Variants {
+			for _, vKey := range variant.Keys {
+				key := vKey
+				if key.String() != "*" {
+					keys = append(keys, key.String())
+				}
+			}
+		}
+
+		rv := res[i]
+
+		matches := matchSelectorKeys(rv, keys)
+		pref = append(pref, matches)
+	}
+
+	return pref
+}
+
+func (e *executer) filterVariants(m ast.Matcher, pref [][]string) []ast.Variant {
+	// Step 3: Filter Variants
+	var filteredVariants []ast.Variant
+
+	for _, variant := range m.Variants {
+		matchesAllSelectors := true
+
+		for i, keyOrder := range pref {
+			key := variant.Keys[i]
+			if key.String() == "*" {
+				continue
+			}
+
+			ks := key.String()
+			if !slices.Contains(keyOrder, ks) {
+				matchesAllSelectors = false
+				break
+			}
+		}
+
+		if matchesAllSelectors {
+			filteredVariants = append(filteredVariants, variant)
+		}
+	}
+
+	return filteredVariants
+}
+
+func (e *executer) sortVariants(filteredVariants []ast.Variant, pref [][]string) []SortableVariant {
+	// Step 4: Sort Variants
+	sortable := make([]SortableVariant, 0, len(filteredVariants))
+
+	for _, variant := range filteredVariants {
+		sortable = append(sortable, SortableVariant{Score: -1, Variant: variant})
+	}
+
+	for i := len(pref) - 1; i >= 0; i-- {
+		matches := pref[i]
+
+		for tupleIndex, tuple := range sortable {
+			key := tuple.Variant.Keys[i]
+			currentScore := len(matches)
+
+			if key.String() != "*" {
+				ks := key.String()
+				if position := findPosition(ks, matches); position != -1 {
+					currentScore = position
+				}
+			}
+
+			sortable[tupleIndex].Score = currentScore
+		}
+
+		SortVariants(sortable)
+	}
+
+	return sortable
+}
+
+func (e *executer) selectBestVariant(sortable []SortableVariant) (strings.Builder, error) { //nolint:unparam
+	// Select the best variant
+	bestVariant := sortable[0].Variant
+
+	var output strings.Builder
+
+	for _, patternElement := range bestVariant.QuotedPattern {
+		if err := e.resolvePattern(patternElement); err != nil {
+			return strings.Builder{}, fmt.Errorf("resolve pattern element: %w", err)
+		}
+	}
+
+	return output, nil
+}
+
+// The SortVariants function.
+func SortVariants(sortable []SortableVariant) {
+	sort.Sort(SortableVariantSlice(sortable))
+}
+
+func matchSelectorKeys(rv Selector, keys []string) []string {
+	var matches []string
+
+	for _, key := range keys {
+		if value, ok := rv.Value.(string); ok {
+			if key == value {
+				matches = append(matches, key)
+			}
+		}
+	}
+
+	return matches
+}
+
+type SortableVariant struct {
+	Variant ast.Variant
+	Score   int
+}
+
+type SortableVariantSlice []SortableVariant
```

Comment:
Avoid using type name in identifier names unless the context requires it.

Insteaf of `Slice`, you can use plural name.

```suggestion
type SortableVariants []SortableVariant
```